### DISCUSSION TITLE
Discovery via Unicast

### DISCUSSION THREAD

**Discussion by amielc1** [OpenDDS/OpenDDS#4654] [Q&A] [answered]
How can I set the discovery process via Unicast instead of Multicast - in `rtps.ini` file? 

I have 2 participants that send messages to the same Topic. 
I need to use unicast addresses, so each participant sends messages to another IP (another transport). 
all transports defined in 'rtps.ini' file. 
I noticed that my application sends messages to other topic subscribers although I didn't specify their address in my rtps file. while I disabled the multicast.

Thank you for your help! 
Amiel.

> [common]
> DCPSGlobalTransportConfig=participant1_transport 
> DCPSDefaultDiscovery=unicast_rtps
> DCPSDebugLevel=10 
> ORBDebugLevel=10
> ORBLogFile=LogFilePublisher.log
> 
> [domain/43]
> DiscoveryConfig=unicast_rtps
> 
> [rtps_discovery/unicast_rtps]
> SedpRtpsRelayAddress=172.198.166.99:8888
> SpdpRtpsRelayAddress=172.198.166.99:8889
> UseRtpsRelay=1 
> 
> [config/participant1_transport]
> transports=udp1
> 
> [config/participant2_transport]
> transports=udp2
> 
> [transport/udp1]
> transport_type=rtps_udp
> use_multicast=0
> local_address=172.198.166.124:5555
> DataRtpsRelayAddress=192.198.166.99:6666  
> UseRtpsRelay=1
> 
> [transport/udp2]
> transport_type=rtps_udp
> use_multicast=0
> local_address=172.198.166.124:3333
> DataRtpsRelayAddress=172.198.166.124:4444  
> UseRtpsRelay=1

**Comment by jrw972** [comment, marked as answer]
You have a couple of options but first some background.

RTPS Discovery assumes multicast for SPDP.
The goal is to discover all of the other participants in the domain.
Discovery will then match readers and writers.
So, a participant might communicate with another participant up to the point that they realize they don't need to communicate further.
This is normal and expected.
If you want to partition your application to avoid messages, then you can look at using different domains, partitions, topics, etc.

Multicast for SEDP can be disabled by setting [SedpMulticast](https://opendds.readthedocs.io/en/latest-release/devguide/run_time_configuration.html#cfg-prop-rtps_discovery-SedpMulticast) to false.
Multicast for the transport can be disabled by setting `use_multicast=0` as you have in the ini file.
So, the remaining issue is working around the SPDP multicast.
In general, there is only one way to completely disable multicast sends which I will discuss below.

First, you can set [SpdpSendAddrs](https://opendds.readthedocs.io/en/latest-release/devguide/run_time_configuration.html#cfg-prop-rtps_discovery-SpdpSendAddrs).
This doesn't disable multicast but it does cause a participant to send SPDP to the specified addresses.
So, if you know the addresses of all the participants up front, you can use this option.

Second, you can use the [multicast repeater](https://github.com/OpenDDS/cloud-multicast-repeater).
The idea here is that multicast is supported within a host but not between hosts like in many cloud environments.
So, the participant sends multicast SPDP to the multicast repeater which is on the same host.
The multicast repeater then sends unicast to multicast repeaters on other hosts.
The multicast repeater receiving the unicast SPDP converts it back to multicast.
 
Third, you can use the RtpsRelay.
In this case, the participants will send their SPDP messages through the RtpsRelay via unicast.
Using the RtpsRelay allows one to disable multicast completely by specifying `RtpsRelayOnly`.
In this case, the multicast messages are suppressed and all messages go to the RtpsRelay.
Use this with caution since it precludes the possibility of participant to participant communication with is more efficient.

**Reply by amielc1** [reply]
Great! 
The best solution for me is - Rtps Relay.
